### ISSUE TITLE
Feature request: Singular-value decomposition (SVD)

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#110]
_From @caisq on January 20, 2018 17:16_

This can mimic tensorflow's o`tf.linalg.svd`: https://www.tensorflow.org/api_docs/python/tf/svd

This is useful for, among other things, the generating of orthogonal initial weights for RNNs.

_Copied from original issue: tensorflow/tfjs-core#571_

**Comment by kedevked** [comment]
Hi, 
I will work on it

**Comment by caisq** [comment]
Thanks, @kedevked Looking forward to your PR. Let us know if you need any help or guidance.

**Comment by kedevked** [comment]
I am going to implement it using the qr method. The algorithm is described [here](http://www.math.tamu.edu/~dallen/linear_algebra/chpt6.pdf). I wonder from which discrepancy can we consider the convergence to be reached for a singular value, meaning that if at the ith iteration we have the value j and at the (i+1)th iteration we have the value k, I want to stop the iteration when k-i < threshold. What should the threshold be ?
Also, please could you review the [solve PR](https://github.com/tensorflow/tfjs-core/pull/1241), as I am going to use it to find the singular vector for each singular value ?

**Comment by stken2050** [comment]
I have a question, for node.

For Python version, svd is already there;
https://www.tensorflow.org/api_docs/python/tf/linalg/svd
and I guess these Python apis are based on C binaries.

tfjs for node is also based on the same C binary, I've heard. So I think it should be trivial to implement all of the apis that are already implemented in Python version, for tfjs for node.

I understand all of the apis of  tfjs on browser is pure js and it seems a long way development to cover basic apis. 

So, here's my question. 

Is there any development policies that apis for both browser and node are synchlonized? If so, even there are exsiting binaries to bind from node, we must wait every api for browser pure js implementation.

If I miss something, plese advice. Thanks.

**Comment by lp74** [comment]
@kedevked, with a shift (-s I) the convergence would be faster.

**Comment by kedevked** [comment]
@nsthorat @caisq can you please tell us if it is still in the roadmap of tfjs to implement linear algebra operators ? If so, maybe you can give some pointers in the direction to go. Thanks !

**Comment by EddieOne** [comment]
It would be nice to have a way to test SVD in TFJS.

**Comment by kedevked** [comment]
@EddieOne Yes indeed, that would be nice to have. But I guess the tfjs team for now have decided not to include this operators in the core library.

**Comment by EddieOne** [comment]
As I understand it, we need more floating points in JS to get the accuracy needed for SVD. But I am fuzzy on the details. I was going to try more interpolation without SVD and see what happens.

**Comment by tsol** [comment]
Not a single one implementation of SVD i found in NPM decomposes [8,9] matrix exactly like pyhons Numpy.lialg.svd function. I really hoped tfjs would have one...

**Comment by gaikwadrahul8** [comment]
Hi, @nsthorat 

Thank you for opening this issue for tracking purposes. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFJs team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TFJs [version](https://github.com/tensorflow/tfjs/releases/tag/tfjs-v4.11.0) with the latest compatible hardware configuration which could potentially resolve the issue. We can keep the issue open if it is still relevant. Please confirm if we need to keep the issue open.

Thank you for your support and cooperation.

**Comment by avmohansai** [comment]
Hello @nsthorat  ,
Thank you for bringing this issue to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by Varun-S10** [comment]
Hi @nsthorat,
Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
